src/util.rs

```diff
@@ -76,6 +107,36 @@ impl std::ops::SubAssign for Timeout {
     }
 }
 
+impl From<Timeout> for Deadline {
+    fn from(t: Timeout) -> Deadline {
+        if let Timeout::After(dur) = t {
+            Deadline::new(Some(cmp::min(Deadline::MAX_FLUSH_DURATION, dur)))
+        } else {
+            Deadline::new(None)
+        }
+    }
+}
+
+impl From<Deadline> for Timeout {
+    fn from(d: Deadline) -> Timeout {
+        if let Deadline::Never = d {
```

Comment:
maybe we could use `impl From<&Deadline> for Timeout` here to not duplicate the logic?